### ISSUE TITLE
Can we use config file to pass so many args to components, instead of bare cmd flags?

### ISSUE THREAD

**Issue by astraw99** [kubernetes/kubernetes#108821]
### What would you like to be added?

Components like kube-apiserver, controller-manager, scheduler, have dozens of options to pass through cmd flags,
which seems clunky.
So can we use a single config file to pass so many args to components, like this:
```
kube-apiserver --config=/path/to/config.yaml
```
instead of currently cmd flags below:
```
kube-apiserver --advertise-address=xxx --allow-privileged=true --authorization-mode=Node,RBAC ...
```

### Why is this needed?

1. Manually config so many cmd flags is clunky and hard to maintain.
2. When debug the component, it is not easy to get all the flags currently in use.

Thus, with a single config file, it is easy to config, maintain and debug.

**Comment by kerthcet** [comment]
At least to kube-scheduler, we have configuration file named `KubeSchedulerConfiguration` , you can refer to this https://kubernetes.io/docs/reference/config-api/kube-scheduler-config.v1beta3/ for details. Can this help you?

**Comment by astraw99** [comment]
@deads2k @liggitt Could you please give some advice here, thanks :)

**Comment by lavalamp** [comment]
There has been at least one attempt to add config files in addition to flags, but AFAIK it stalled out before it got to apiserver.

The issue is it's fiddly to do something like this while not breaking any current consumers of flags.

I'm not sure we'd say no to someone wanting to try to add this, but it's not high on the priority list?

**Comment by aojea** [comment]
> There has been at least one attempt to add config files in addition to flags, but AFAIK it stalled out before it got to apiserver.
> 
> The issue is it's fiddly to do something like this while not breaking any current consumers of flags.
> 
> I'm not sure we'd say no to someone wanting to try to add this, but it's not high on the priority list?

kube-proxy moved to config , overriding flags, and is a source of problems , mainly because configuration files can not leverage the downward API for automat per node configurations
https://github.com/kubernetes/kubernetes/issues/108737

**Comment by thockin** [comment]
I recall discussing this with @mtaufen YEARS ago, when this effort kicked off.

I still think it makes sense that flags should override config files, but ISTR some weird corner cases where that was wrong.  I can not recall the cases.  The reason for my belief is that it seems common to want to distribute configs by file-distribution (configmap, ansible, git, etc) but still have locally-defined values in flags.

As @lavalamp said - we'd not say no to someone wanting to try to finish this work.

**Comment by astraw99** [comment]
As the [Configuration as Code](https://www.cloudbees.com/blog/configuration-as-code-everything-need-know) said, maybe we can consider to use config files to pass default args, and keep using cmd flags to do the merge and override.

How do you guys think of using the mixture of them?

**Comment by liggitt** [comment]
The componentconfig APIs was what https://github.com/kubernetes/community/tree/master/archive/wg-component-standard spend a fair amount of time trying to sketch out.

Previous discussions on config/flag merging or not, or mixing instance-specific config:
* https://github.com/kubernetes/kubernetes/pull/101273#issuecomment-823357862
* https://github.com/kubernetes/enhancements/pull/1439/files

cc @deads2k @sttts as well

**Comment by mtaufen** [comment]
> I still think it makes sense that flags should override config files, but ISTR some weird corner cases where that was wrong.

For Kubelet I think this was basically always the behavior we wanted. I think kube-proxy (memory is fuzzy though) tried to make config and flags mutually exclusive to avoid the backwards-compatibility requirements around supporting both simultaneously. Then dynamic kubelet config added its own layer of complexity to the flag precedence implementation (but now that's being removed by @SergeyKanzhelev , so yay simplifying!), but didn't fundamentally change the need to implement precedence for backwards compatibility.

I wrote a guide on how to add existing kubelet flags to config files a few years ago, I think most of it should still be valid: https://docs.google.com/document/d/10Kz__miCWEOekC6WnsUdxYWSLrf-t1AkHHpZN57ei5c/edit#heading=h.mupqgfuyjwnm

> kube-proxy moved to config , overriding flags, and is a source of problems , mainly because configuration files can not leverage the downward API for automat per node configurations

The concern around lack of downward API for in-cluster workloads (like kube-proxy-as-daemonset) that want to adopt config files instead of flags has been brought up before, and it's a friction point. You can do it with env vars and sed and such but it's a cumbersome approach. @neolit123 had proposed some alternatives a few years ago: https://docs.google.com/document/d/1Dvct469xfjkgy3tjWMAKvRAJo4CmGH4cgSVGTDpay6A/edit 

I imagine it's a friction point for many workloads that want to inject things into config files, have we seen any clean solutions emerge in other parts of the ecosystem?

> I'm not sure we'd say no to someone wanting to try to add this, but it's not high on the priority list?

That's pretty much it. AIUI it still comes up in SIG planning meetings every so often, but always gets backlogged behind other high priority work. The original plan was to try it with kubelet and kube-proxy first (and we were driving mostly in SIG-Node in concert with the dynamic kubelet config idea), but as @lavalamp mentioned we stalled out before making it anywhere close to kube-apiserver.

**Comment by mtaufen** [comment]
It is likely a good idea to make flags and config mutually exclusive. I'd probably recommend it for new attempts at this. I think I've said that before on at least one other thread :P. 

With kubelet I was trying to make it possible to migrate flags incrementally, because there were tons of flags and tons of open questions about which ones we should keep in config and which we should put on a deprecation path, so it didn't seem trivial to just move them all at once, and that's where the precedence requirement came from. If I had to do it again, I think I'd have deferred dynamic kubelet config and all discussions around deprecation of existing options, made flags and config mutually exclusive, duplicated all the flags in the config format in one go, deprecated flags as the config mechanism, and once that deprecation was finished, _then_ start the conversation around which options should be in a v2 config.

Original flag precedence issue: https://github.com/kubernetes/kubernetes/issues/56171

**Comment by mtaufen** [comment]
Some history of kubelet and config here: https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/edit#

**Comment by lavalamp** [comment]
> lack of downward API for in-cluster workloads

You can use an initContainer to plumb env vars into a config file, if you wanted -- is that the solution people are calling too hacky?

**Comment by mtaufen** [comment]
> You can use an initContainer to plumb env vars into a config file, if you wanted -- is that the solution people are calling too hacky?

Yeah that sounds right. People wanted something more direct/built in.
